import os
import sys
import pathlib
import io

# --- Streamlit/UI ---
import streamlit as st
import pandas as pd

# --- gRPC ---
import grpc

# Asegura que Python encuentre los stubs generados en ml/
ROOT = pathlib.Path(__file__).resolve().parents[1]
sys.path.append(str(ROOT / "ml"))

try:
    import sentiment_pb2 as pb
    import sentiment_pb2_grpc as pb_grpc
    GRPC_AVAILABLE = True
except ImportError:
    GRPC_AVAILABLE = False

# --------- Cliente gRPC ----------
def make_stub(addr: str) -> 'pb_grpc.SentimentServiceStub':
    """
    Crea y devuelve el stub del servicio gRPC.
    """
    channel = grpc.insecure_channel(addr)
    return pb_grpc.SentimentServiceStub(channel)

def ping(stub: 'pb_grpc.SentimentServiceStub') -> str:
    """
    Verifica salud del servicio. Devuelve el status.
    """
    return stub.Ping(pb.PingRequest()).status

def predict_text(stub: 'pb_grpc.SentimentServiceStub', text: str) -> tuple[str, float]:
    """
    Env√≠a un texto y obtiene (label, score).
    """
    resp = stub.Predict(pb.PredictRequest(text=text))
    return resp.label, resp.score

def predict_batch(
    stub: 'pb_grpc.SentimentServiceStub', texts: list[str], chunk: int = 128
) -> list[tuple[str, float]]:
    """
    Predicci√≥n en lote con particionado para no exceder tama√±o de mensaje.
    """
    out: list[tuple[str, float]] = []
    for i in range(0, len(texts), chunk):
        part = texts[i : i + chunk]
        resp = stub.PredictBatch(pb.PredictBatchRequest(texts=part))
        out.extend(list(zip(resp.labels, resp.scores)))
    return out

def summarize_text(
    stub: 'pb_grpc.SentimentServiceStub', text: str, sentences: int = 3
) -> list[str]:
    """
    Solicita resumen extractivo LexRank. Devuelve oraciones.
    """
    resp = stub.Summarize(pb.SummarizeRequest(text=text, sentences=sentences))
    return list(resp.sentences)

# --------- Base de datos simulada de rese√±as ----------
if 'reviews_db' not in st.session_state:
    st.session_state.reviews_db = []

def save_review(text: str, recommend: bool, sentiment_label: str = None, sentiment_score: float = None):
    """
    Guarda una rese√±a en la base de datos simulada.
    """
    review = {
        'id': len(st.session_state.reviews_db) + 1,
        'texto': text,
        'recomienda': recommend,
        'sentiment_label': sentiment_label,
        'sentiment_score': sentiment_score,
        'timestamp': pd.Timestamp.now()
    }
    st.session_state.reviews_db.append(review)

# --------- CSS Personalizado ----------
def apply_custom_css():
    st.markdown("""
    <style>
        /* Fondo beige para toda la aplicaci√≥n */
        .stApp {
            background-color: #C73D36;
        }
        
        /* Espaciado y contenedor principal */
        .block-container {
            padding-top: 2rem !important;
            padding-bottom: 2rem !important;
            max-width: 700px !important;
        }
        
        /* Ocultar elementos de streamlit por defecto */
        #MainMenu {visibility: hidden;}
        footer {visibility: hidden;}
        header {visibility: hidden;}
        
        /* T√≠tulo principal */
        .main-title {
            color: white !important;
            font-size: 28px !important;
            font-weight: 600 !important;
            margin-bottom: 1.5rem !important;
            text-align: center !important;
            padding: 1rem 0 !important;
        }
        
        /* Subt√≠tulos de tabs */
        .tab-subheader {
            color: white !important;
            font-size: 22px !important;
            font-weight: 500 !important;
            margin-bottom: 1rem !important;
            text-align: left !important;
        }
        
        /* Secci√≥n de recomendaci√≥n */
        .recommend-container {
            background-color: rgba(255, 255, 255, 0.1);
            padding: 1rem;
            border-radius: 10px;
            margin: 1rem 0;
            border: 1px solid rgba(255, 255, 255, 0.2);
        }
        
        .recommend-text {
            color: white !important;
            font-size: 16px !important;
            font-weight: 500 !important;
            margin: 0 !important;
        }
        
        /* √Årea de texto personalizada */
        .stTextArea > div > div > textarea {
            background-color: #3e4651 !important;
            color: white !important;
            border: 1px solid rgba(255, 255, 255, 0.3) !important;
            border-radius: 8px !important;
            font-size: 16px !important;
            padding: 1rem !important;
            min-height: 120px !important;
        }
        
        .stTextArea > div > div > textarea::placeholder {
            color: rgba(255, 255, 255, 0.6) !important;
        }
        
        /* Botones */
        .stButton > button {
            background-color: #4a90e2 !important;
            color: white !important;
            border: none !important;
            padding: 0.8rem 2rem !important;
            border-radius: 8px !important;
            font-size: 16px !important;
            font-weight: 600 !important;
            width: 100% !important;
            margin-top: 1rem !important;
            transition: all 0.3s ease !important;
            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2) !important;
        }
        
        .stButton > button:hover {
            background-color: #357abd !important;
            transform: translateY(-1px) !important;
            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.3) !important;
        }
        
        /* Mensajes de estado */
        .stSuccess {
            background-color: rgba(40, 167, 69, 0.2) !important;
            color: white !important;
            border: 1px solid rgba(40, 167, 69, 0.4) !important;
        }
        
        .stError {
            background-color: rgba(220, 53, 69, 0.2) !important;
            color: white !important;
            border: 1px solid rgba(220, 53, 69, 0.4) !important;
        }
        
        .stInfo {
            background-color: rgba(74, 144, 226, 0.2) !important;
            color: white !important;
            border: 1px solid rgba(74, 144, 226, 0.4) !important;
        }
        
        .stWarning {
            background-color: rgba(255, 193, 7, 0.2) !important;
            color: white !important;
            border: 1px solid rgba(255, 193, 7, 0.4) !important;
        }
        
        /* DataFrames */
        .stDataFrame {
            background-color: rgba(255, 255, 255, 0.1) !important;
        }
        
        /* Tabs */
        .stTabs [data-baseweb="tab-list"] {
            gap: 8px;
        }
        
        .stTabs [data-baseweb="tab"] {
            background-color: rgba(255, 255, 255, 0.1) !important;
            color: white !important;
            border-radius: 8px 8px 0 0 !important;
        }
        
        /* Separadores */
        hr {
            border-color: rgba(255, 255, 255, 0.3) !important;
            margin: 1.5rem 0 !important;
        }
        
        /* Texto general */
        p, div, span, li, label {
            color: white !important;
        }
        
        /* M√©tricas */
        .metric-container {
            background-color: rgba(255, 255, 255, 0.1);
            padding: 1rem;
            border-radius: 8px;
            margin: 0.5rem 0;
            border: 1px solid rgba(255, 255, 255, 0.2);
        }
        
        /* Footer */
        .footer-text {
            text-align: center;
            color: rgba(255, 255, 255, 0.8) !important;
            font-size: 14px !important;
            margin-top: 1.5rem !important;
        }
    </style>
    """, unsafe_allow_html=True)

# --------- UI Functions ----------
def ui_write_review():
    """
    Interfaz para escribir rese√±as con an√°lisis de sentimientos.
    """
    st.markdown('<h2 class="tab-subheader">üìù Escribir Rese√±a</h2>', unsafe_allow_html=True)
    
    # Secci√≥n de recomendaci√≥n a amigos
    st.markdown('<div class="recommend-container">', unsafe_allow_html=True)
    col1, col2 = st.columns([3.5, 1])
    
    with col1:
        st.markdown('<p class="recommend-text">¬øRecomendar√≠as a amigos?</p>', unsafe_allow_html=True)
    
    with col2:
        recommend_friends = st.toggle("", value=False, key="recommend_toggle")
    
    st.markdown('</div>', unsafe_allow_html=True)
    
    # √Årea de texto para la rese√±a
    review_text = st.text_area(
        "",
        placeholder="Escribe tu rese√±a aqu√≠...",
        height=120,
        key="review_textarea",
        label_visibility="collapsed"
    )
    
    # Bot√≥n de env√≠o
    if st.button("üì§ Enviar Rese√±a", type="primary", use_container_width=True):
        if review_text.strip():
            # An√°lisis de sentimientos si gRPC est√° disponible
            sentiment_label = None
            sentiment_score = None
            
            if GRPC_AVAILABLE:
                try:
                    addr = os.getenv("APP_GRPC_ADDR", "localhost:50051")
                    stub = make_stub(addr)
                    sentiment_label, sentiment_score = predict_text(stub, review_text.strip())
                except Exception as e:
                    st.warning(f"‚ö† An√°lisis de sentimientos no disponible: {e}")
            
            # Guardar rese√±a
            save_review(review_text.strip(), recommend_friends, sentiment_label, sentiment_score)
            
            # Mostrar resultados
            st.success("‚úÖ ¬°Rese√±a enviada exitosamente!")
            
            # Mostrar resumen
            st.markdown("---")
            st.markdown("üìã Resumen de tu rese√±a:")
            
            col1, col2, col3 = st.columns(3)
            with col1:
                recommend_text = "‚úÖ S√≠" if recommend_friends else "‚ùå No"
                st.info(f"*Recomendar√≠as:* {recommend_text}")
            with col2:
                st.info(f"*Caracteres:* {len(review_text)}")
            with col3:
                if sentiment_label and sentiment_score is not None:
                    sentiment_emoji = "üòä" if sentiment_label.lower() == "positive" else "üòî" if sentiment_label.lower() == "negative" else "üòê"
                    st.info(f"*Sentimiento:* {sentiment_emoji} {sentiment_label} ({sentiment_score:.2f})")
                else:
                    st.info("*Sentimiento:* No disponible")
            
            st.markdown("*Tu rese√±a:*")
            st.markdown(f'"{review_text}"')
            
            # Mensaje de agradecimiento
            st.success("¬°Gracias por tu feedback! üíô")
            
        else:
            st.error("‚ö† Por favor, escribe una rese√±a antes de enviar.")

def ui_sentiment_analysis():
    """
    Interfaz para an√°lisis individual de sentimientos.
    """
    st.markdown('<h2 class="tab-subheader">üß† An√°lisis de Sentimientos</h2>', unsafe_allow_html=True)
    
    if not GRPC_AVAILABLE:
        st.error("‚ùå Servicio de an√°lisis de sentimientos no disponible. Instala las dependencias gRPC.")
        return
    
    txt = st.text_area("Texto para analizar:", height=160, placeholder="Pega aqu√≠ el texto a analizar...")
    
    if st.button("üîç Analizar Sentimiento", type="primary", use_container_width=True, disabled=not txt.strip()):
        try:
            addr = os.getenv("APP_GRPC_ADDR", "localhost:50051")
            stub = make_stub(addr)
            label, score = predict_text(stub, txt.strip())
            
            # Mostrar resultado con colores
            sentiment_emoji = "üòä" if label.lower() == "positive" else "üòî" if label.lower() == "negative" else "üòê"
            sentiment_color = "success" if label.lower() == "positive" else "error" if label.lower() == "negative" else "info"
            
            if sentiment_color == "success":
                st.success(f"{sentiment_emoji} *Sentimiento:* {label} | *Confianza:* {score:.3f}")
            elif sentiment_color == "error":
                st.error(f"{sentiment_emoji} *Sentimiento:* {label} | *Confianza:* {score:.3f}")
            else:
                st.info(f"{sentiment_emoji} *Sentimiento:* {label} | *Confianza:* {score:.3f}")
                
        except Exception as e:
            st.error(f"‚ùå Error conectando con el servicio: {e}")

def ui_reviews_database():
    """
    Interfaz para ver todas las rese√±as guardadas.
    """
    st.markdown('<h2 class="tab-subheader">üìä Base de Datos de Rese√±as</h2>', unsafe_allow_html=True)
    
    if not st.session_state.reviews_db:
        st.info("üìù No hay rese√±as guardadas a√∫n.")
        return
    
    # Convertir a DataFrame
    df = pd.DataFrame(st.session_state.reviews_db)
    
    # M√©tricas generales
    st.markdown("üìà Estad√≠sticas Generales:")
    col1, col2, col3, col4 = st.columns(4)
    
    with col1:
        st.markdown('<div class="metric-container">', unsafe_allow_html=True)
        st.metric("Total Rese√±as", len(df))
        st.markdown('</div>', unsafe_allow_html=True)
    
    with col2:
        recommend_count = df['recomienda'].sum()
        st.markdown('<div class="metric-container">', unsafe_allow_html=True)
        st.metric("Recomiendan", recommend_count)
        st.markdown('</div>', unsafe_allow_html=True)
    
    with col3:
        if 'sentiment_label' in df.columns and df['sentiment_label'].notna().any():
            positive_count = (df['sentiment_label'] == 'positive').sum()
            st.markdown('<div class="metric-container">', unsafe_allow_html=True)
            st.metric("Positivas", positive_count)
            st.markdown('</div>', unsafe_allow_html=True)
        else:
            st.markdown('<div class="metric-container">', unsafe_allow_html=True)
            st.metric("Sentimiento", "N/A")
            st.markdown('</div>', unsafe_allow_html=True)
    
    with col4:
        avg_length = df['texto'].str.len().mean()
        st.markdown('<div class="metric-container">', unsafe_allow_html=True)
        st.metric("Longitud Prom.", f"{avg_length:.0f} chars")
        st.markdown('</div>', unsafe_allow_html=True)
    
    # Mostrar tabla de rese√±as
    st.markdown("---")
    st.markdown("üìã Todas las Rese√±as:")
    
    # Preparar DataFrame para mostrar
    display_df = df.copy()
    display_df['recomienda'] = display_df['recomienda'].map({True: '‚úÖ S√≠', False: '‚ùå No'})
    if 'sentiment_label' in display_df.columns:
        display_df['sentiment_label'] = display_df['sentiment_label'].fillna('N/A')
        display_df['sentiment_score'] = display_df['sentiment_score'].fillna(0).round(3)
    
    st.dataframe(display_df, use_container_width=True)
    
    # Bot√≥n para limpiar base de datos
    if st.button("üóë Limpiar Base de Datos", type="secondary"):
        st.session_state.reviews_db = []
        st.success("‚úÖ Base de datos limpiada.")
        st.experimental_rerun()

def main():
    """
    Funci√≥n principal que construye la interfaz completa.
    """
    # Configuraci√≥n de p√°gina
    st.set_page_config(
        page_title="Sistema de Rese√±as con IA",
        page_icon="üçΩ",
        layout="wide"
    )
    
    # Aplicar CSS personalizado
    apply_custom_css()
    
    # T√≠tulo principal
    st.markdown('<h1 class="main-title">üçΩ Sistema de Rese√±as de Restaurante con IA</h1>', unsafe_allow_html=True)
    
    # Verificar conexi√≥n gRPC si est√° disponible
    if GRPC_AVAILABLE:
        try:
            addr = os.getenv("APP_GRPC_ADDR", "localhost:50051")
            stub = make_stub(addr)
            status = ping(stub)
            st.toast(f"ü§ñ IA conectada: {status}", icon="‚úÖ")
        except Exception as e:
            st.toast(f"‚ö† IA no disponible: {str(e)[:50]}...", icon="‚ö†")
    
    # Pesta√±as principales
    tab1, tab2, tab3 = st.tabs(["üìù Escribir Rese√±a", "üß† An√°lisis IA", "üìä Base de Datos"])
    
    with tab1:
        ui_write_review()
    
    with tab2:
        ui_sentiment_analysis()
    
    with tab3:
        ui_reviews_database()
    
    # Footer
    st.markdown("---")
    st.markdown(
        """
        <div class="footer-text">
            <p>üçΩ Sistema de Rese√±as con An√°lisis de Sentimientos | Powered by BETO + gRPC ü§ñ</p>
        </div>
        """, 
        unsafe_allow_html=True
    )

if __name__ == "__main__":
    main()